File: cli/python/ros2bag_converter/mcap.py
# SPDX-License-Identifier: Apache-2.0
"""MCAP (.mcap) reader for ROS 2 bags.

MCAP is ROS 2's current default rosbag2 storage format. Messages are the same
CDR payloads as the SQLite format, and schemas (ros2msg) are embedded — so this
module only parses the MCAP container and reuses the shared CDR decoder.

Uncompressed chunks (and chunk-less files) need only the standard library.
zstd / lz4 chunk compression needs the optional `zstandard` / `lz4` packages;
a clear error is raised if such a chunk is encountered without the library.
"""

from __future__ import annotations

import io
import struct

from . import decoder
from .bag import Topic

MAGIC = b"\x89MCAP0\r\n"

# record opcodes
_HEADER, _FOOTER, _SCHEMA, _CHANNEL, _MESSAGE, _CHUNK, _DATA_END = (
    0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x0F,
)


class _Reader:
    __slots__ = ("b", "o")

    def __init__(self, buf, off=0):
        self.b = buf
        self.o = off

    def u8(self):
        v = self.b[self.o]; self.o += 1; return v

    def u16(self):
        v = struct.unpack_from("<H", self.b, self.o)[0]; self.o += 2; return v

    def u32(self):
        v = struct.unpack_from("<I", self.b, self.o)[0]; self.o += 4; return v

    def u64(self):
        v = struct.unpack_from("<Q", self.b, self.o)[0]; self.o += 8; return v

    def s(self):
        n = self.u32(); v = self.b[self.o:self.o + n].decode("utf-8", "replace"); self.o += n; return v

    def bytes_u32(self):
        n = self.u32(); v = self.b[self.o:self.o + n]; self.o += n; return v


def _iter_records(buf, start, end):
    o = start
    while o + 9 <= end:
        op = buf[o]
        ln = struct.unpack_from("<Q", buf, o + 1)[0]
        cstart = o + 9
        content = buf[cstart:cstart + ln]
        o = cstart + ln
        yield op, content
        if op == _FOOTER:
            break


def _decompress(comp, data, uncompressed_size):
    if not comp:
        return data
    if comp == "zstd":
        try:
            import zstandard
        except ImportError:
            raise RuntimeError(
                "MCAP chunk uses zstd compression; install 'zstandard' (pip install zstandard) to read it."
            )
        return zstandard.ZstdDecompressor().stream_reader(io.BytesIO(data)).read()
    if comp == "lz4":
        try:
            import lz4.frame as lz4frame
        except ImportError:
            raise RuntimeError(
                "MCAP chunk uses lz4 compression; install 'lz4' (pip install lz4) to read it."
            )
        return lz4frame.decompress(data)
    raise RuntimeError(f"unsupported MCAP chunk compression: {comp!r}")


class McapBag:
    """Duck-typed like bag.Bag: .topics, .total, .gmin/.gmax/.storage/.distro,
    .each_message(topic_id, max_rows), .close()."""

    def __init__(self):
        self.topics = []
        self.total = 0
        self.gmin = None
        self.gmax = None
        self.storage = "mcap"
        self.distro = ""
        self._msgs = {}  # channel_id -> list[(log_time, data)]
        self._truncated = False

    def each_message(self, topic_id, max_rows=None):
        data = sorted(self._msgs.get(topic_id, []), key=lambda x: x[0])
        self._truncated = False
        for i, (log_time, blob) in enumerate(data):
            if max_rows is not None and i >= max_rows:
                self._truncated = True
                break
            yield log_time, blob

    def close(self):
        pass


def open_mcap(path: str, metadata_path: "str | None" = None) -> McapBag:
    with open(path, "rb") as fh:
        buf = fh.read()
    if buf[:8] != MAGIC:
        raise ValueError("Not an MCAP file — bad magic bytes.")

    schemas = {}   # id -> name
    channels = {}  # id -> (topic, schema_id, msg_encoding)
    bag = McapBag()

    def handle(op, content):
        r = _Reader(content)
        if op == _SCHEMA:
            sid = r.u16(); name = r.s(); enc = r.s(); data = r.bytes_u32()
            schemas[sid] = name
            if enc == "ros2msg" and data:
                try:
                    decoder.register_encoded_definition(decoder.norm(name), data.decode("utf-8", "replace"))
                except Exception:
                    pass
        elif op == _CHANNEL:
            cid = r.u16(); schema_id = r.u16(); topic = r.s(); msg_enc = r.s()
            channels[cid] = (topic, schema_id, msg_enc)
            bag._msgs.setdefault(cid, [])
        elif op == _MESSAGE:
            cid = r.u16(); r.u32(); log_time = r.u64(); r.u64()
            bag._msgs.setdefault(cid, []).append((log_time, content[r.o:]))
        elif op == _CHUNK:
            r.u64(); r.u64(); usize = r.u64(); r.u32(); comp = r.s()
            n = r.u64()                      # records byte-length (uint64)
            records = r.b[r.o:r.o + n]
            raw = _decompress(comp, records, usize)
            for iop, icontent in _iter_records(raw, 0, len(raw)):
                handle(iop, icontent)

    for op, content in _iter_records(buf, 8, len(buf)):
        handle(op, content)

    for cid in sorted(channels, key=lambda c: channels[c][0]):
        topic, schema_id, msg_enc = channels[cid]
        type_ = schemas.get(schema_id) or "(unknown)"
        mlist = bag._msgs.get(cid, [])
        cnt = len(mlist)
        tmin = min((t for t, _ in mlist), default=None)
        tmax = max((t for t, _ in mlist), default=None)
        bag.topics.append(Topic(cid, topic, type_, msg_enc or "cdr", cnt, tmin, tmax,
                                decoder.is_decodable(type_)))
        bag.total += cnt
        if cnt > 0:
            bag.gmin = tmin if bag.gmin is None else min(bag.gmin, tmin)
            bag.gmax = tmax if bag.gmax is None else max(bag.gmax, tmax)

    if metadata_path:
        import os
        import re
        if os.path.exists(metadata_path):
            with open(metadata_path, "r", encoding="utf-8", errors="replace") as fh:
                meta = fh.read()
            dm = re.search(r"ros_distro:\s*(\S+)", meta)
            if dm:
                bag.distro = dm.group(1)
    return bag

File: cli/node/src/mcap.js
// SPDX-License-Identifier: Apache-2.0
'use strict';
/* ============================================================================
   MCAP (.mcap) reader — ROS 2's current default rosbag2 format. Messages are
   the same CDR payloads with schemas embedded, so only the container parser is
   new; decoding reuses ./decoder. Returns the uniform Bag interface used by
   ./bag (topics, info, eachMessage, close).

   Uncompressed chunks need nothing extra. zstd chunks use Node's built-in zstd
   (Node ≥ 22.15); lz4 is not supported here — use the .db3 export or Python CLI.
   ========================================================================== */

const fs = require('fs');
const zlib = require('zlib');
const { norm, registerEncodedDefinition, registry } = require('./decoder');

const MAGIC = Buffer.from([0x89, 0x4d, 0x43, 0x41, 0x50, 0x30, 0x0d, 0x0a]);

function decompress(comp, data) {
  if (!comp || comp.length === 0) return data;
  if (comp === 'zstd') {
    if (typeof zlib.zstdDecompressSync === 'function') return zlib.zstdDecompressSync(data);
    throw new Error('MCAP chunk uses zstd compression; needs Node ≥ 22.15 (built-in zstd), or convert the bag to .db3 / use the Python CLI.');
  }
  if (comp === 'lz4') {
    throw new Error("MCAP chunk uses lz4 compression, which this Node reader can't decompress; use the Python CLI with the 'lz4' package, or convert to .db3.");
  }
  throw new Error(`unsupported MCAP chunk compression: ${comp}`);
}

function iterRecords(buf, handle) {
  let o = 0; const end = buf.length;
  while (o + 9 <= end) {
    const op = buf.readUInt8(o);
    const ln = Number(buf.readBigUInt64LE(o + 1));
    const cstart = o + 9;
    const content = buf.subarray(cstart, cstart + ln);
    o = cstart + ln;
    handle(op, content);
    if (op === 0x02) break;   // footer — last record before trailing magic
  }
}

function openMcap(path, metadataPath) {
  const buf = fs.readFileSync(path);
  if (!buf.subarray(0, 8).equals(MAGIC)) throw new Error('Not an MCAP file — bad magic bytes.');

  const schemas = new Map();   // id -> name
  const channels = new Map();  // id -> { topic, schemaId, msgEnc }
  const msgs = new Map();      // id -> [[logTime(BigInt), data(Buffer)]]

  function handle(op, c) {
    let o = 0;
    if (op === 0x03) {                    // Schema
      const sid = c.readUInt16LE(o); o += 2;
      let n = c.readUInt32LE(o); o += 4; const name = c.toString('utf8', o, o + n); o += n;
      n = c.readUInt32LE(o); o += 4; const enc = c.toString('utf8', o, o + n); o += n;
      n = c.readUInt32LE(o); o += 4; const data = c.subarray(o, o + n); o += n;
      schemas.set(sid, name);
      if (enc === 'ros2msg' && data.length) {
        try { registerEncodedDefinition(norm(name), data.toString('utf8')); } catch (e) { /* one bad def */ }
      }
    } else if (op === 0x04) {             // Channel
      const cid = c.readUInt16LE(o); o += 2; const schemaId = c.readUInt16LE(o); o += 2;
      let n = c.readUInt32LE(o); o += 4; const topic = c.toString('utf8', o, o + n); o += n;
      n = c.readUInt32LE(o); o += 4; const msgEnc = c.toString('utf8', o, o + n); o += n;
      channels.set(cid, { topic, schemaId, msgEnc });
      if (!msgs.has(cid)) msgs.set(cid, []);
    } else if (op === 0x05) {             // Message
      const cid = c.readUInt16LE(o); o += 2; o += 4; /* seq */
      const logTime = c.readBigUInt64LE(o); o += 8; o += 8; /* publish_time */
      if (!msgs.has(cid)) msgs.set(cid, []);
      msgs.get(cid).push([logTime, c.subarray(o)]);
    } else if (op === 0x06) {             // Chunk
      o += 8; o += 8; o += 8; o += 4;     // start, end, uncompressed_size, crc
      let n = c.readUInt32LE(o); o += 4; const comp = c.toString('utf8', o, o + n); o += n;
      const recLen = Number(c.readBigUInt64LE(o)); o += 8;
      const raw = decompress(comp, c.subarray(o, o + recLen));
      iterRecords(raw, handle);
    }
  }

  iterRecords(buf.subarray(8), handle);

  const topics = [];
  let total = 0, gMin = null, gMax = null;
  const ids = [...channels.keys()].sort((a, b) => (channels.get(a).topic < channels.get(b).topic ? -1 : 1));
  for (const cid of ids) {
    const { topic, schemaId, msgEnc } = channels.get(cid);
    const type = schemas.get(schemaId) || '(unknown)';
    const list = msgs.get(cid) || [];
    let tmin = null, tmax = null;
    for (const [lt] of list) { if (tmin === null || lt < tmin) tmin = lt; if (tmax === null || lt > tmax) tmax = lt; }
    topics.push({
      id: cid, name: topic, type, fmt: msgEnc || 'cdr', cnt: list.length,
      tmin: tmin === null ? null : Number(tmin), tmax: tmax === null ? null : Number(tmax),
      decodable: registry.has(norm(type)),
    });
    total += list.length;
    if (list.length > 0) {
      const a = Number(tmin), b = Number(tmax);
      if (gMin === null || a < gMin) gMin = a;
      if (gMax === null || b > gMax) gMax = b;
    }
  }

  let distro = '';
  if (metadataPath && fs.existsSync(metadataPath)) {
    const m = fs.readFileSync(metadataPath, 'utf8'); const dm = m.match(/ros_distro:\s*(\S+)/); if (dm) distro = dm[1];
  }

  return {
    topics,
    info: { total, gMin, gMax, storage: 'mcap', distro },
    eachMessage(topicId, cb, maxRows) {
      const list = (msgs.get(topicId) || []).slice().sort((x, y) => (x[0] < y[0] ? -1 : x[0] > y[0] ? 1 : 0));
      let n = 0, truncated = false;
      for (const [lt, data] of list) {
        if (maxRows != null && n >= maxRows) { truncated = true; break; }
        cb(lt.toString(), data);   // ts as string → full-precision, like the sqlite path
        n++;
      }
      return { count: n, truncated };
    },
    close() { /* nothing to release */ },
  };
}

module.exports = { openMcap };
